src/mailmap.c

```diff
@@ -0,0 +1,381 @@
+		return false;
+	}
+
+	before->len = off - range.p;
+	after->p = off + 1;
+	after->len = (range.p + range.len) - after->p;
+	return true;
+}
+
+/* Trim whitespace from the beginning and end of the range */
+static void range_trim(char_range *range) {
+	while (range->len > 0 && git__isspace(range->p[0])) {
+		++range->p;
+		--range->len;
+	}
+	while (range->len > 0 && git__isspace(range->p[range->len - 1]))
+		--range->len;
+}
+
+/**
+ * If `buf` is not NULL, copies range into it with a '\0', and bumps buf.
+ * If `size` is not NULL, adds the number of bytes to be written to it.
+ * returns a pointer to the copied string, or NULL.
+ */
+static const char *range_copyz(char **buf, size_t *size, char_range src)
+{
+	char *s = NULL;
+	if (src.p == NULL)
+		return NULL;
+
+	if (size)
+		*size += src.len + 1;
+
+	if (buf) {
+		s = *buf;
+		memcpy(s, src.p, src.len);
+		s[src.len] = '\0';
+		*buf += src.len + 1;
+	}
+	return s;
+}
+
+struct git_mailmap {
+	git_vector entries;
+};
+
+/**
+ * Parse a single entry out of a mailmap file.
+ * Advances the `file` range past the parsed entry.
+ */
+static int git_mailmap_parse_single(
+	char_range *file,
+	bool *found,
+	char_range *real_name,
+	char_range *real_email,
+	char_range *replace_name,
+	char_range *replace_email)
+{
+	char_range line, comment, name_a, email_a, name_b, email_b;
+	bool two_emails = false;
+
+	*found = false;
+	*real_name = NULL_RANGE;
+	*real_email = NULL_RANGE;
+	*replace_name = NULL_RANGE;
+	*replace_email = NULL_RANGE;
+
+	while (file->len > 0) {
+		/* Get the line, and remove any comments */
+		range_split(*file, '\n', &line, file);
+		range_split(line, '#', &line, &comment);
+
+		/* Skip blank lines */
+		range_trim(&line);
+		if (line.len == 0)
+			continue;
+
+		/* Get the first name and email */
+		if (!range_split(line, '<', &name_a, &line))
+			return -1; /* garbage in line */
+		if (!range_split(line, '>', &email_a, &line))
+			return -1; /* unfinished <> pair */
+
+		/* Get an optional second name and/or email */
+		two_emails = range_split(line, '<', &name_b, &line);
+		if (two_emails && !range_split(line, '>', &email_b, &line))
+			return -1; /* unfinished <> pair */
+
+		/* Trim whitespace from around names */
+		range_trim(&name_a);
+		range_trim(&name_b);
+
+		*found = true;
+		if (name_a.len > 0)
+			*real_name = name_a;
+
+		if (two_emails) {
+			if (email_a.len > 0)
+				*real_email = email_a;
+			*replace_email = email_b;
+
+			if (name_b.len > 0)
+				*replace_name = name_b;
+		} else {
+			*replace_email = email_a;
+		}
+		break;
+	}
+
+	return 0;
+}
+
+int git_mailmap_parse(
+	git_mailmap **mailmap,
+	const char *data,
+	size_t size)
+{
+	char_range file = { data, size };
+	git_mailmap_entry* entry = NULL;
+	int error = 0;
+
+	if (memchr(data, '\0', size) != NULL)
+		return -1; /* data may not contain '\0's */
+
+	*mailmap = git__calloc(1, sizeof(git_mailmap));
+	if (!*mailmap)
+		return -1;
+
+	/* XXX: Is it worth it to precompute the size? */
```

Comment:
Probably not. We're always expanding vectors by a factor of 1.5 every time, so it's not like every entry will cause a new allocation